internal/controller/monovertexrollout_controller.go

```diff
@@ -0,0 +1,361 @@
+
+	if monoVertexChildResourceReason == "Progressing" {
+		rollout.Status.MarkChildResourcesUnhealthy("Progressing", "MonoVertex Progressing", rollout.Generation)
+	} else if monoVertexPhase == numaflowv1.MonoVertexPhaseFailed || monoVertexChildResourceStatus == "False" {
+		rollout.Status.MarkChildResourcesUnhealthy("MonoVertexFailed", "MonoVertex Failed", rollout.Generation)
+	} else if monoVertexPhase == numaflowv1.MonoVertexPhaseUnknown || monoVertexChildResourceStatus == "Unknown" {
+		rollout.Status.MarkChildResourcesHealthUnknown("MonoVertexUnkown", "MonoVertex Phase Unknown", rollout.Generation)
+	} else {
+		rollout.Status.MarkChildResourcesHealthy(rollout.Generation)
+	}
+
+}
+
+func (r *MonoVertexRolloutReconciler) needsUpdate(old, new *apiv1.MonoVertexRollout) bool {
+	if old == nil {
+		return true
+	}
+	if !equality.Semantic.DeepEqual(old.Finalizers, new.Finalizers) {
+		return true
+	}
+	return false
+}
+
+func (r *MonoVertexRolloutReconciler) updateMonoVertex(ctx context.Context, monoVertexRollout *apiv1.MonoVertexRollout, newMonoVertexDef *kubernetes.GenericObject) error {
+	err := kubernetes.UpdateCR(ctx, r.restConfig, newMonoVertexDef, "monovertices")
+	if err != nil {
+		return err
+	}
+
+	monoVertexRollout.Status.MarkDeployed(monoVertexRollout.Generation)
+	return nil
+}
+
+func (r *MonoVertexRolloutReconciler) updateMonoVertexRolloutStatus(ctx context.Context, monoVertexRollout *apiv1.MonoVertexRollout) error {
+	rawSpec := runtime.RawExtension{}
+	err := util.StructToStruct(&monoVertexRollout.Spec, &rawSpec)
+	if err != nil {
+		return fmt.Errorf("unable to convert MonoVertexRollout Spec to GenericObject Spec: %v", err)
+	}
+
+	rawStatus := runtime.RawExtension{}
+	err = util.StructToStruct(&monoVertexRollout.Status, &rawStatus)
+	if err != nil {
+		return fmt.Errorf("unable to convert MonoVertexRollout Status to GenericObject Status: %v", err)
+	}
+
+	obj := kubernetes.GenericObject{
+		TypeMeta:   monoVertexRollout.TypeMeta,
+		ObjectMeta: monoVertexRollout.ObjectMeta,
+		Spec:       rawSpec,
+		Status:     rawStatus,
+	}
+
+	return kubernetes.UpdateStatus(ctx, r.restConfig, &obj, "monovertexrollouts")
+}
+
+func (r *MonoVertexRolloutReconciler) updateMonoVertexRolloutStatusToFailed(ctx context.Context, monoVertexRollout *apiv1.MonoVertexRollout, err error) error {
+	monoVertexRollout.Status.MarkFailed(err.Error())
+	return r.updateMonoVertexRolloutStatus(ctx, monoVertexRollout)
+}
+
+func (r *MonoVertexRolloutReconciler) ErrorHandler(monoVertexRollout *apiv1.MonoVertexRollout, err error, reason, msg string) {
+	// r.customMetrics.MonoVerticesSyncFailed.WithLabelValues().Inc()
+	r.recorder.Eventf(monoVertexRollout, corev1.EventTypeWarning, reason, msg+" %v", err.Error())
+}
+
+func getMonoVertexChildResourceHealth(conditions []metav1.Condition) (metav1.ConditionStatus, string) {
+	for _, cond := range conditions {
+		switch cond.Type {
+		case "DaemonHealthy", "PodsHealthy":
```

Comment:
You confirmed there are only two conditions for monovertex? Thanks!

Reply:
Yes, these are the only conditions for subresource health. The other condition a MonoVertex will have is `Deployed`.